### DISCUSSION TITLE
Concurrency Test Released, prior tests updated

### DISCUSSION THREAD

**Discussion by TANSixu** [SUSTech-CS305-Fall22/CS305-Project-Skeleton#33] [Announcements]
PR #32 <del>Info will come later.</del>
<del>[update] May have to delay the description, got the fking covid, sorry :( </del>

[update] I am back, sorry for the delay

Please run the following command to fetch the updated changes:
```
git pull staff
```
If you did not configure your local git following the [tutorial](https://github.com/SUSTech-CS305-Fall22/CS305-Project-Skeleton), you can update your local files according to the [file change record](https://github.com/SUSTech-CS305-Fall22/CS305-Project-Skeleton/pull/32/files)

# Concurrency Test
This test examines the basic function of your concurrency.
Peer1 will be downloading chunk2,3 from Peer2, 3 concurrently.

This test will be running with network simulator, with topology test/tmp3/topo3.map. You can assume there will be no packet loss in this test. The script will check if you can correctly download all chunks, and it will generate a concurrency_analysis plot, which will be checked on the presentation.

### Without concurrency
![non-concurrency](https://user-images.githubusercontent.com/66773369/209672369-7322d0f1-5fa3-4d43-aa3a-a04e0ae797d9.png)
### With concurrency
![concurrency](https://user-images.githubusercontent.com/66773369/209672467-b4c8be01-d3cd-4ef9-ab3b-c15ebb1298eb.png)

You can also visualize your concurrency without running the test:
```
python3 test/concurrency_visulizer.py <log_file>
```

# Updata
## simsocket.py
* More detailed log. Simsocket now generates log with more details like checkersock, including pkt_type, seq, and ack. Now a log line will be like:
```
2022-12-23 19:27:12,613 -+- PEER1_LOGGER -+- DEBUG -+- Receiving a type3 pkt from ('127.0.0.1', 48003) via spiffy, seq469, ack0, pkt_len1016
```
* User log. An "add_log" method is provided to help users debug in log. This method allows users to print any message into log file.
```
def add_log(self, msg)
``` 

## grader.py
We found that previous testing scripts may have abnormal behavior on platforms other than the provided ones. Problems are mainly introduced by
1. Some systems start a new process more slowly, so peers are not fully started in testing scripts.
2. Multiprocessing/threading issue

To address these issues, we
1. Modified testing scripts, now they should be 99% the same as running in command line.
2. Provided commands of tests. You can see commands similar to the following on the top of testing files:
```
This test is equivalent to run (except for packet loss):
In shell1:
python3 src/peer.py -p test/tmp2/nodes2.map -c test/tmp2/data1.fragment -m 1 -i 1 -t 60

In shell2:
python3 src/peer.py -p test/tmp2/nodes2.map -c test/tmp2/data2.fragment -m 1 -i 2 -t 60

In shell1:
DOWNLOAD test/tmp2/download_target.chunkhash test/tmp2/download_result.fragment
```
You can run them in shells to debug (stdout and stderr will be printed if you run in shells)

**Comment by Pilusimida** [comment]
学长多多注意休息，早日康复！！👍🏻

**Reply by TANSixu** [reply]
感谢！
